gecode/flatzinc/lastval.hh

```diff
@@ -0,0 +1,183 @@
+/* -*- mode: C++; c-basic-offset: 2; indent-tabs-mode: nil -*- */
+/*
+ *  Main authors:
+ *     Jip J. Dekker <jip.dekker@monash.edu>
+ *
+ *  Copyright:
+ *     Jip J. Dekker, 2023
+ *
+ *  This file is part of Gecode, the generic constraint
+ *  development environment:
+ *     http://www.gecode.org
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining
+ *  a copy of this software and associated documentation files (the
+ *  "Software"), to deal in the Software without restriction, including
+ *  without limitation the rights to use, copy, modify, merge, publish,
+ *  distribute, sublicense, and/or sell copies of the Software, and to
+ *  permit persons to whom the Software is furnished to do so, subject to
+ *  the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be
+ *  included in all copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ *  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ *  MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ *  LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ *  OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ *  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifndef __FLATZINC_LASTVAL_HH__
+#define __FLATZINC_LASTVAL_HH__
+
+#include <gecode/flatzinc.hh>
+#include <gecode/int.hh>
+
+using namespace Gecode::Int;
+
+namespace Gecode { namespace FlatZinc {
+
+  class LastValInt : public UnaryPropagator<IntView, PC_INT_VAL> {
```

Comment:
Is there a particular reason (except perhaps simplicity) that there are individual recording propagators for each variable?

In https://github.com/Gecode/gecode/tree/feature/on_restart2_nary_lastval_int I added a variant that uses a single nary propagator for integer variables as an example of how it can be done. It should be a small improvement in both performance and memory, as a single propagator is smaller and takes less time to copy. The single linear scan through the array for recording should be insignificant in comparison.

Reply:
Note that I have only compiled the above code, not tested it at all. It is just a sketch on how it could look, not a drop-in replacement that I know works.

Reply:
That is actually a very good idea. I'll see when I have some time to make those changes

Reply:
Probably worth at least a couple of measurements as well, just to ensure it behaves as expected.

Reply:
I've updated the `LastVal` propagator to use an N-ary propagator, and revisiting this allowed me to clean it up in general. I've now changed it to use templates to remove much of the boilerplate code.

I'll have a look at adding the MiniZinc tests that I have as FlatZinc tests in the coming days.

Reply:
I saw that the subscription condition is PC_INT_VAL, that should be templated as well. 

Alternatively, you could use [PC_GEN_ASSIGNED](https://www.gecode.org/doc-latest/reference/group__TaskVarMEPC.html#ga2e782baaa2ef69ccd53b88e04d5510fb)